tango/step_info.py

```diff
@@ -1,13 +1,73 @@
-from collections import OrderedDict
-from dataclasses import asdict, dataclass
+import getpass
+import logging
+import os
+import platform
+import socket
+import sys
+from dataclasses import dataclass, field
 from datetime import datetime, timedelta
 from enum import Enum
-from typing import Any, Dict, Optional, Set
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Set, Tuple
 
 import pytz
 
-from .common.util import local_timezone
+from .common.from_params import FromParams
+from .common.util import jsonify, local_timezone, replace_steps_with_unique_id
 from .step import Step
+from .version import VERSION
+
+logger = logging.getLogger(__name__)
+
+
+def get_pip_packages() -> Optional[List[Tuple[str, str]]]:
+    """
+    Get the current working set of pip packages. Equivalent to running ``pip freeze``.
+    """
+    # Adapted from the Weights & Biases client library:
+    # github.com/wandb/client/blob/a04722575eee72eece7eef0419d0cea20940f9fe/wandb/sdk/internal/meta.py#L56-L72
+    try:
+        import pkg_resources
+
+        return sorted([(d.key, d.version) for d in iter(pkg_resources.working_set)])
+    except Exception as exc:
+        logger.exception("Error saving pip packages: %s", exc)
+    return None
+
+
+def get_conda_env() -> Optional[Dict[str, Any]]:
+    """
+    Get the current conda environment. Equivalent to running ``conda env export``.
+    """
+    # Adapted from the Weights & Biases client library:
+    # github.com/wandb/client/blob/a04722575eee72eece7eef0419d0cea20940f9fe/wandb/sdk/internal/meta.py#L74-L87
+    current_shell_is_conda = os.path.exists(os.path.join(sys.prefix, "conda-meta"))
+    if current_shell_is_conda:
+        import subprocess
+
+        import yaml
+
+        try:
+            result = subprocess.run(["conda", "env", "export"], capture_output=True)
+            if (
+                result.returncode != 0
+                and result.stderr is not None
+                and "Unable to determine environment" in result.stderr.decode()
+            ):
+                result = subprocess.run(
+                    ["conda", "env", "export", "-n", "base"], capture_output=True
+                )
+
+            if result.returncode != 0:
+                if result.stderr is not None:
+                    logger.exception("Error saving conda packages: %s", result.stderr.decode())
+                else:
+                    result.check_returncode()
+            elif result.stdout is not None:
+                return yaml.safe_load(result.stdout.decode())
+        except Exception as exc:
+            logger.exception("Error saving conda packages: %s", exc)
+    return None
```

Comment:
What do you think about getting rid of this entire thing? I know in principle it would be quite useful to have for debugging. But in practice we almost never look at it, and it takes several seconds per step to do this. When the steps are fast, this is where 90% of the execution time is.

Reply:
That's fine with me. It seems too brittle anyway.